### ISSUE TITLE
Partitioned stream workers

### ISSUE THREAD

**Issue by palsivertsen** [nats-io/jetstream#464]
In IoT environments there's often a need to process lots and lots of sensor data from devices and one can group similar devices into JetStream streams, with a common configuration for retention etc, where each device has it's own unique subject within the stream. Sometimes it's useful to preserve the order of messages during processing, e.g looking for threshold breaches. This is isn't a problem when having a single exclusive consumer, but things get tricky if you want to scale consumers, e.g in response to a temporary burst in messages. Adding or removing consumers would require all consumers to agree on who should process which subjects.

To summarize I want to:
* Dynamically add and remove workers (consumers?) for processing a stream.
* Balance subjects among workers where each subject is processed by at most one worker at the time.
* At least once delivery semantics in case of a rebalance.

This is very similar to #103, but with a more relaxed constraint on message ordering. I.e. when a message arrives for processing it's guaranteed that all messages up until this point has arrived in order and that this message is the next in line. What's not guaranteed is if the message has been processed before (at least once delivery).

**Comment by ripienaar** [comment]
This is something we'd look at after initial GA but its on our radar for sure

**Comment by ripienaar** [comment]
Closing as JS is now GA, we have this on roadmap
